src/server/middleware/clientinfo/clientinfo.go

```diff
@@ -0,0 +1,86 @@
+// Copyright Project Harbor Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package clientinfo
+
+import (
+	"net/http"
+	"regexp"
+	"strings"
+
+	"github.com/goharbor/harbor/src/lib/config"
+
+	"github.com/goharbor/harbor/src/lib"
+	"github.com/goharbor/harbor/src/server/middleware"
+)
+
+var (
+	// De-facto standard header keys.
+	xForwardedFor = http.CanonicalHeaderKey("X-Forwarded-For")
+	xRealIP       = http.CanonicalHeaderKey("X-Real-IP")
+
+	// RFC7239 defines a new "Forwarded: " header designed to replace the
+	// existing use of X-Forwarded-* headers.
+	// e.g. Forwarded: for=192.0.2.60;proto=https;by=203.0.113.43
+	forwarded = http.CanonicalHeaderKey("Forwarded")
+	// Allows for a sub-match of the first value after 'for=' to the next
+	// comma, semi-colon or space. The match is case-insensitive.
+	forRegex = regexp.MustCompile(`(?i)(?:for=)([^(;|,| )]+)`)
+)
+
+func getIP(r *http.Request) string {
```

Comment:
if client request through multi-level proxy server(e.g. `client-->proxy 1-->proxy 2->harbor` ), what's the function will return ?

Reply:
The function returns output based on the HTTP headers present in the request:
- XFF (X-Forwarded-For): This here will return the first IP (usually* the client’s IP).
- X-Real-IP: This will return the IP as set by the last proxy. It could be either the client’s IP or the last proxy’s IP, depending on how the proxies are configured.
- Forwarded: This here will capture the first "for=" entry, which is usually* the client’s IP.

I’ve used "usually" because these values can be easily altered by proxies in the middle. To prevent this, we should have a way to define which proxies we trust. This issue has already been raised [here](https://github.com/goharbor/harbor/issues/20367#issuecomment-2095575188) .

Hope that answers your question.